### DISCUSSION TITLE
how to use meshopt encoder to encode THREE.scene by js

### DISCUSSION THREAD

**Discussion by Frankling** [donmccurdy/glTF-Transform#512] [Q&A] [answered]
how to use meshopt encoder to encode THREE.scene by js

**Comment by donmccurdy** [comment, marked as answer]
You'll need to convert your three.js scene to a glTF file before using glTF Transform to optimize it, probably using [THREE.GLTFExporter](https://threejs.org/docs/?q=gltfex#examples/en/exporters/GLTFExporter). I'd recommend testing the glTF file in online viewers like https://gltf.report/ after doing so, to make sure the file contains everything you expect.

**Reply by Frankling** [reply]
Thanks for reply, but when i use the gltfExporter the size is larger than its origial file ,so is there any way to encode it by js,

**Reply by donmccurdy** [reply]
Exporting a scene with GLTFExporter is definitely a lossy process. You can re-optimize it back to roughly its original size using tools like glTF Transform or gltfpack, but this will have side effects:

- Geometry: Methods like Draco or Meshopt are lossy and applying them repeatedly may increase that loss.
- Images: Re-compressing JPEG image compression will increase loss.

Because of those issues, it would be a bad idea to repeatedly import and re-export the same GLB file with GLTFExporter. If there's any way your application can avoid that, I would. Better solutions are possible if you want to render to three.js out of a lossless glTF wrapper (see https://github.com/donmccurdy/glTF-Transform-Render/) but I haven't finished my own implementation of that, and can't really give you a ready-to-use solution there.

***

If you still want to go ahead with re-optimizing the glTF file after exporting with GLTFExporter, I would look at the example here: https://github.com/mrdoob/three.js/pull/22227#issuecomment-897310955. The same process would work with Meshopt compression if you follow the [meshopt documentation ](https://gltf-transform.donmccurdy.com/classes/extensions.meshoptcompression.html). Note that if the cause of the size increase with GLTFExporter is the _textures_ (this is very common) then you would need texture compression instead, which this library does yet currently offer (see https://github.com/donmccurdy/glTF-Transform/pull/506).
